lib/puppet/configurer.rb

```diff
@@ -418,6 +426,40 @@ def save_last_run_summary(report)
     Puppet.log_exception(detail, _("Could not save last run local report: %{detail}") % { detail: detail })
   end
 
+  # Submit updated facts to the Puppet Server
+  #
+  # This method will clear all current fact values, load a fresh set of
+  # fact data, and then submit it to the Puppet Server.
+  #
+  # @return [true] If fact submission succeeds.
+  # @return [false] If an exception is raised during fact generation or
+  #   submission.
+  def resubmit_facts
+    ::Facter.clear
+    facts = find_facts
+
+    saved_fact_terminus = Puppet::Node::Facts.indirection.terminus_class
+    begin
+      Puppet::Node::Facts.indirection.terminus_class = :rest
+
+      server = Puppet::Node::Facts::Rest.server
+      Puppet.notice(_("Uploading facts for '%{node}' to: '%{server}'") % {
+                    node: facts.name,
+                    server: server})
+
+      Puppet::Node::Facts.indirection.save(facts, nil, :environment => Puppet::Node::Environment.remote(@environment))
```

Comment:
We might want to double check other places `Puppet::Node::Facts.indirection.save` is being set to make sure all the required query items are being set. @joshcooper notes it's not always obvious by looking at the indirector what is required.

Reply:
This section of the code was copied wholesale from the `puppet facts upload` command:

https://github.com/puppetlabs/puppet/blob/6.8.1/lib/puppet/face/facts.rb#L78-L84

Reply:
As far as I know, `puppet facts upload` is the only other user of this API for remote submission as the "common" path for facts to go to the Server is via the Catalog terminus instead of the Facts terminus.

Reply:
Sorry this is fine, and `puppet facts upload` too. I was confused because the compiler specifies a [`transaction_uuid`](https://github.com/puppetlabs/puppet/blob/ca796215b95825f22b0ea30f56a27af181eda53a/lib/puppet/indirector/catalog/compiler.rb#L42) when saving the facts. But the [puppetdb terminus ignores it](https://github.com/puppetlabs/puppetdb/blob/e2156116dbf60d0c2212526fb92b76a56432c7be/puppet/lib/puppet/indirector/facts/puppetdb.rb#L34-L43).